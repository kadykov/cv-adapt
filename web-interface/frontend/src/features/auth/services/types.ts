export interface LoginRequest {
  email: string;
  password: string;
}

export interface RegisterRequest extends LoginRequest {
  personal_info?: {
    first_name?: string;
    last_name?: string;
  };
}

export interface ValidationResponse {
  valid: boolean;
  user?: {
    id: number;
    email: string;
    created_at: string;
    personal_info?: {
      first_name?: string;
      last_name?: string;
    } | null;
  };
  error?: string;
}
