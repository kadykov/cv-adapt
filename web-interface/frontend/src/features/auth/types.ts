export interface User {
  id: number;
  email: string;
  created_at: string;
  personal_info?: Record<string, never> | null;
}

export interface AuthResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  user: User;
}

export interface LoginRequest {
  username: string;
  password: string;
}

export interface RegisterRequest {
  email: string;
  password: string;
}
