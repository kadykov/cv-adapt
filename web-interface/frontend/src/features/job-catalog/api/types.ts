export interface JobDescriptionCreate {
  title: string;
  description: string;
  language_code: string;
}

export interface JobDescriptionUpdate {
  title?: string | null;
  description?: string | null;
}

export interface JobDescriptionResponse {
  id: number;
  title: string;
  description: string;
  language_code: string;
  created_at: string;
  updated_at: string | null;
}

export type JobsResponse = JobDescriptionResponse[];

export interface ErrorResponse {
  detail: {
    message: string;
    code?: string;
    field?: string;
  };
}
