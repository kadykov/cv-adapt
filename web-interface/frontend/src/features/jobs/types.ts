import type { components } from '../../lib/api/types';

export type Job = components['schemas']['JobDescriptionResponse'];
export type JobCreate = components['schemas']['JobDescriptionCreate'];
export type JobUpdate = components['schemas']['JobDescriptionUpdate'];
