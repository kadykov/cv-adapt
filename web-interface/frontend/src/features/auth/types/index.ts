export interface User {
  id: number;
  email: string;
  personal_info: Record<string, unknown> | null;
  created_at: string;
}

export interface LoginCredentials {
  email: string;
  password: string;
  remember?: boolean;
}

export interface RegistrationData {
  email: string;
  password: string;
  acceptTerms: boolean;
}

export interface AuthResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  user: User;
}

export interface AuthState {
  user: User | null;
  token: string | null;
  isAuthenticated: boolean;
  isLoading: boolean;
}
