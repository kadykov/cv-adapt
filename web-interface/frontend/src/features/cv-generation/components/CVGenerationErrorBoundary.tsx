import { Component, ReactNode } from 'react';
import { Button } from '@headlessui/react';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
  error?: Error;
}

/**
 * Error boundary specific to CV generation components
 * Provides fallback UI for runtime errors
 */
export class CVGenerationErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error };
  }

  handleReset = () => {
    this.setState({ hasError: false, error: undefined });
  };

  render() {
    if (this.state.hasError) {
      return (
        <div
          className="alert alert-error"
          role="alert"
          aria-live="assertive"
        >
          <div
            className="flex flex-col items-start gap-2"
          >
            <h3
              className="font-semibold"
              id="error-heading"
            >
              Failed to render CV generation component
            </h3>
            <p
              className="text-sm"
              id="error-message"
              aria-describedby="error-heading"
            >
              {this.state.error?.message || 'An unknown error occurred'}
            </p>
            <Button
              onClick={this.handleReset}
              className="btn btn-sm btn-ghost mt-2"
              type="button"
              aria-label="Try loading the component again"
            >
              Try Again
            </Button>
          </div>
        </div>
      );
    }

    return this.props.children;
  }
}
