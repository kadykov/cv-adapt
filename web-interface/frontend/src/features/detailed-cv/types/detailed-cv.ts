/**
 * Types for the Detailed CV feature
 * Leveraging generated types from OpenAPI schema
 */
import {
  DetailedCVCreate,
  DetailedCVResponse,
} from '../../../lib/api/generated-types';
import { LanguageCode } from '../../../lib/language/types';

export type { DetailedCVCreate, DetailedCVResponse };

/**
 * Form data for creating/editing a detailed CV
 */
export interface DetailedCVFormData {
  language_code: LanguageCode;
  content: string;
  is_primary: boolean;
}

/**
 * Mapping function to convert form data to API request
 */
export function mapFormToApiRequest(
  formData: DetailedCVFormData,
): Omit<DetailedCVCreate, 'content'> & { content: { markdown: string } } {
  return {
    language_code: formData.language_code,
    content: { markdown: formData.content }, // Store markdown content in a structured way
    is_primary: formData.is_primary,
  };
}

/**
 * Mapping function to convert API response to form data
 */
export function mapApiToFormData(
  response: DetailedCVResponse,
): DetailedCVFormData {
  return {
    language_code: response.language_code as LanguageCode,
    content:
      typeof response.content === 'object' &&
      response.content !== null &&
      'markdown' in response.content
        ? String(response.content.markdown)
        : '',
    is_primary: response.is_primary,
  };
}
