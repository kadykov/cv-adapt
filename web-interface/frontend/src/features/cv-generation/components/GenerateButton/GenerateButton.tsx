import { Button } from '@headlessui/react';
import { Icon } from '@iconify/react';
import { CVGenerationErrorBoundary } from '../CVGenerationErrorBoundary';

interface GenerateButtonProps {
  /**
   * ID of the job to generate CV for
   */
  jobId: number;


  /**
   * Optional class name for styling
   * @default 'btn btn-primary gap-2'
   */
  className?: string;

  /**
   * Optional disabled state
   * @default false
   */
  disabled?: boolean;

  /**
   * Click handler
   */
  onClick?: () => void;
}

/**
 * Button component that initiates CV generation for a specific job.
 */
export function GenerateButton({
  jobId,
  className = 'btn btn-primary gap-2',
  disabled = false,
  onClick,
}: GenerateButtonProps) {
  const handleClick = () => {
    if (!disabled && onClick) {
      onClick();
    }
  };

  return (
    <CVGenerationErrorBoundary>
      <Button
        onClick={handleClick}
        className={className}
        disabled={disabled}
        type="button"
        aria-label={`Generate CV for job ${jobId}`}
        aria-disabled={disabled}
      >
        <Icon
          icon="heroicons:document"
          className="w-5 h-5"
          aria-hidden="true"
          data-testid="generate-cv-icon"
        />
        <span>Generate CV</span>
      </Button>
    </CVGenerationErrorBoundary>
  );
}
