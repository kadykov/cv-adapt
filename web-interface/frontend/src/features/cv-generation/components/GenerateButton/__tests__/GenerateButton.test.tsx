import { describe, test, expect, vi, afterEach } from 'vitest';
import { screen, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { GenerateButton } from '../GenerateButton';
import { CVGenerationErrorBoundary } from '../../CVGenerationErrorBoundary';
import { renderWithRouter } from '../../../../../lib/test/render-utils';
import { LanguageCode } from '@/lib/language/types';

/**
 * GenerateButton Component Tests
 *
 * Testing Strategy:
 * - Component behavior tests (rendering, click handling)
 * - Accessibility compliance
 * - Error boundary functionality
 * - Language code validation
 *
 * Note: Navigation logic is tested at the integration level
 * in generation-flow.integration.test.tsx
 */

// Mock Iconify's Icon component for testing
vi.mock('@iconify/react', () => ({
  Icon: ({
    'data-testid': testId,
    className,
    icon,
  }: {
    'data-testid'?: string;
    className?: string;
    icon: string;
  }) => (
    <svg
      data-testid={testId}
      className={className}
      aria-hidden="true"
      data-icon={icon}
    />
  ),
}));

describe('GenerateButton', () => {
  const defaultProps = {
    jobId: 123,
    onClick: vi.fn(),
    language: LanguageCode.ENGLISH,
  };

  beforeEach(() => {
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.restoreAllMocks();
  });

  describe('rendering', () => {
    test('renders with default props', () => {
      renderWithRouter(<GenerateButton {...defaultProps} />);

      const button = screen.getByRole('button', { name: /generate cv/i });
      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();
    });

    test('renders with custom className', () => {
      const customClass = 'custom-button btn-secondary';
      renderWithRouter(
        <GenerateButton {...defaultProps} className={customClass} />,
      );

      const button = screen.getByRole('button', { name: /generate cv/i });
      expect(button).toHaveClass(customClass);
    });

    test('handles disabled state', () => {
      renderWithRouter(<GenerateButton {...defaultProps} disabled />);

      const button = screen.getByRole('button', { name: /generate cv/i });
      expect(button).toBeDisabled();
      expect(button).toHaveAttribute('aria-disabled', 'true');
    });

    test.each(Object.values(LanguageCode))(
      'renders with valid language code: %s',
      (language) => {
        renderWithRouter(
          <GenerateButton {...defaultProps} language={language} />,
        );
        expect(screen.getByRole('button')).toBeInTheDocument();
      },
    );

    test('handles invalid language code', () => {
      const consoleError = vi.spyOn(console, 'error');
      const invalidLanguage = 'invalid' as LanguageCode;

      renderWithRouter(
        <GenerateButton {...defaultProps} language={invalidLanguage} />,
      );

      expect(screen.queryByRole('button')).not.toBeInTheDocument();
      expect(consoleError).toHaveBeenCalledWith(
        expect.stringContaining('Invalid language code'),
      );
    });
  });

  describe('click handling', () => {
    test('triggers onClick when clicked', async () => {
      const user = userEvent.setup();

      renderWithRouter(<GenerateButton {...defaultProps} />);

      await user.click(screen.getByRole('button', { name: /generate cv/i }));

      expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
    });

    test('does not trigger onClick when disabled', async () => {
      const user = userEvent.setup();

      renderWithRouter(<GenerateButton {...defaultProps} disabled />);

      await user.click(screen.getByRole('button', { name: /generate cv/i }));

      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
  });

  describe('accessibility', () => {
    test('has proper ARIA attributes', () => {
      renderWithRouter(<GenerateButton {...defaultProps} />);

      const button = screen.getByRole('button');
      expect(button).toHaveAttribute(
        'aria-label',
        `Generate CV for job ${defaultProps.jobId}`,
      );
      expect(button).toHaveAttribute('type', 'button');
    });

    test('icon is properly hidden from screen readers', () => {
      renderWithRouter(<GenerateButton {...defaultProps} />);

      const svg = screen.getByTestId('generate-cv-icon');
      expect(svg).toHaveAttribute('aria-hidden', 'true');
      expect(svg.tagName.toLowerCase()).toBe('svg');
      expect(svg).toHaveClass('w-5', 'h-5');
    });
  });

  describe('error handling', () => {
    test('shows error boundary when rendering fails', () => {
      const ThrowError = () => {
        throw new Error('Button render failed');
      };

      renderWithRouter(
        <CVGenerationErrorBoundary>
          <ThrowError />
        </CVGenerationErrorBoundary>,
      );

      expect(screen.getByRole('alert')).toBeInTheDocument();
      expect(screen.getByText(/failed to render/i)).toBeInTheDocument();
    });

    test('recovers from error state', async () => {
      const user = userEvent.setup();

      // Component that throws on first render but not on rerender
      let shouldThrow = true;
      const ToggleError = () => {
        if (shouldThrow) {
          throw new Error('Test error');
        }
        return <div>Recovered content</div>;
      };

      renderWithRouter(
        <CVGenerationErrorBoundary>
          <ToggleError />
        </CVGenerationErrorBoundary>,
      );

      // Verify initial error state
      await waitFor(() => {
        expect(screen.getByRole('alert')).toBeInTheDocument();
        expect(screen.getByText(/failed to render/i)).toBeInTheDocument();
      });

      // Click recovery button
      const retryButton = screen.getByRole('button', { name: /try.*again/i });
      shouldThrow = false;
      await user.click(retryButton);

      // Verify recovery
      await waitFor(() => {
        expect(screen.queryByRole('alert')).not.toBeInTheDocument();
        expect(screen.getByText('Recovered content')).toBeInTheDocument();
      });
    });
  });
});
