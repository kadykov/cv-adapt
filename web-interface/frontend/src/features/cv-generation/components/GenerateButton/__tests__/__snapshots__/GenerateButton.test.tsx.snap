// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GenerateButton > rendering > renders with default props 1`] = `
<button
  aria-disabled="false"
  aria-label="Generate CV for job 123"
  class="btn btn-primary gap-2"
  data-headlessui-state=""
  type="button"
>
  <svg
    aria-hidden="true"
    class="w-5 h-5"
    data-icon="heroicons:document"
    data-testid="generate-cv-icon"
  />
  <span>
    Generate CV
  </span>
</button>
`;
